### PR TITLE
Add hostname sanity checks to prevent using POD name

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

Try to detect if the agent is running in a container with a UTS namespace which is not the root one.
In those cases, the output of `hostname -f` doesn’t reflect the identity of the host and shouldn’t be used.

### Motivation

In some very specific circonstances the agent might need to guess the `hostname` by just doing a `hostname -f` command inside a container.
Those circonstances are:
— the agent hasn’t access to the metadata server of the IaaS,
— the agent hasn’t access to the docker socket,
— the hostname constructed by the agent by concatenating the node name with the cluster name is not valid.

When the agent is running in a POD which has `hostNetwork: false` or, generally speaking, when the agent is running in a container with its dedicated UTS namespace, it shouldn’t use the hostname of the container because it doesn’t reflect the identity of the host and can change at each agent deployment.

### Additional Notes

Without #4492 and with an invalid cluster name (with an underscore),
**Before** this change:
```
$ kubectl get pods
NAME                                          READY   STATUS    RESTARTS   AGE
datadog-48qvg                                 3/3     Running   0          107s

$ kubectl exec datadog-48qvg -c agent agent hostname
Error: ValidHostname: gke-lenaic-pool-1-1c8b424c-spk1-lenaic_gke is not RFC1123 compliant
datadog-48qvg
```

**After** this change:
```
$ kubectl get pod
NAME                                          READY   STATUS             RESTARTS   AGE
datadog-5v98k                                 2/3     Error              0          9s

$ kubectl logs datadog-5v98k -c agent
[…]
2019-11-26 15:43:43 UTC | CORE | DEBUG | (pkg/util/hostname.go:181 in GetHostname) | agent is not running in the root UTS namespace. `hostname -f` cannot be used to detect hostname.
[…]
2019-11-26 15:43:43 UTC | CORE | INFO | (pkg/util/kubernetes/clustername/clustername.go:52 in getClusterName) | Got cluster name lenaic_gke from config
2019-11-26 15:43:43 UTC | CORE | ERROR | (pkg/util/hostname.go:61 in ValidHostname) | ValidHostname: gke-lenaic-pool-1-1c8b424c-wnkn-lenaic_gke is not RFC1123 compliant
[…]
2019-11-26 15:43:43 UTC | CORE | ERROR | (cmd/agent/app/run.go:204 in StartAgent) | Error while getting hostname, exiting: unable to reliably determine the host name. You can define one in the agent config file or in your hosts file
```

### CODE REVIEW

**Review by hkaj — COMMENTED** [review]
that change makes sense, i have one concern though: what if `isHostnameUsable` is false and we're not on ec2? Can we ever end up with no hostname? I'm not sure what's best between a hostname that is actually the pod/container name, and no hostname at all (maybe @DataDog/agent-core can chime in?)

**Comment by olivielpeau** [comment]
FWIW, in terms of user experience, as long as the hostname is unique, it's generally a net benefit to have a hostname vs having no hostname, even if the hostname changes across Agent restarts.

The reasons are that having a hostname, even if it's not stable:

* at least allows users to take advantage of some nice features around having hosts (for example: host shows up on infra list/host map, and all the tagged data -- metrics, etc -- from the host are enriched with the host tags), and
* differentiates metrics coming from one host from metrics coming from a different host if they share the same metric name and tags.

I don't how much of this ^ is actually useful on these containerized environments though.

In terms of our internal constraints, changing the hostname across restarts creates context churn, having no hostname does not. And there may be billing consequences to consider as well, especially if this is a change we introduce for existing customers.

**Comment by L3n41c** [comment]
According to my tests, if the agent ends with no valid hostname, it logs an error and exit.
In kubernetes context, the POD is in `ERROR` state.
So, such a situation cannot really be unseen and ignored. The end-user would still have the possibility to set an explicit hostname via the config. So, we cannot “block” him.

The goal of this change was just to act as a safety belt to prevent that an unstable POD name is used as hostname.

**Review by hkaj — DISMISSED** [review]
lgtm, but let's wait on agent-core's review

**Review by olivielpeau — CHANGES_REQUESTED** [review]
I've only looked at `pkg/util/hostname.go` since I don't really have enough context to review the container-specific logic, left a few comments there but overall LGTM 👍

**Comment by olivielpeau** [comment]
(also, please add a release note on this: an `upgrade` note is welcome since this change will affect some users)

File: pkg/util/containers/types.go
Comment (clamoriniere) [inline pkg/util/containers/types.go]: Defining a type for UTS mode can be interesting.
```
type UTSMode string
const (
	DefaultUTSMode  UTSMode = ""
	HostUTSMode             = "host"
	UnknownUTSMode          = "unknown"
)
```

File: pkg/util/docker/docker_util.go
Comment (clamoriniere) [inline pkg/util/docker/docker_util.go]: I think this function can also be private.

File: pkg/util/docker/global.go
Comment (clamoriniere) [inline pkg/util/docker/global.go]: 👍

File: pkg/util/hostname.go
Comment (hkaj) [inline pkg/util/hostname.go]:
```suggestion
	// https://github.com/kubernetes/kubernetes/blob/cf16e4988f58a5b816385898271e70c3346b9651/pkg/kubelet/dockershim/security_context.go#L203-L205
```

Comment (hkaj) [reply pkg/util/hostname.go]: pinning to a commit instead of master to avoid pointing to something irrelevant as soon as this file gets updated upstream

File: releasenotes/notes/check_hostname_is_usable-dc4b99e6f7242b34.yaml
Comment (hkaj) [inline releasenotes/notes/check_hostname_is_usable-dc4b99e6f7242b34.yaml]:
```suggestion
    Container ID or kubernetes POD names are not suitable for that purpose.
```

Comment (hkaj) [reply releasenotes/notes/check_hostname_is_usable-dc4b99e6f7242b34.yaml]: i guess that's what you meant? In the sense that we shouldn't use container id / pod name to represent the host.

File: pkg/util/hostname.go
Comment (L3n41c) [reply pkg/util/hostname.go]: I’ve add an explicit call to `config.IsContainerized()` in front of all this logic.

File: pkg/util/hostname.go
Comment (L3n41c) [reply pkg/util/hostname.go]: Good catch !!
I've added a stub to handle the case where `kubelet` is not in the build tags. (See `isAgentKubeHostNetwork()`.

File: pkg/util/hostname.go
Comment (L3n41c) [reply pkg/util/hostname.go]: :+1:
I’ve extracted the logic in `func isOSHostnameUsable()`.

File: pkg/util/containers/types.go
Comment (L3n41c) [reply pkg/util/containers/types.go]: Done!

File: pkg/util/docker/docker_util.go
Comment (L3n41c) [reply pkg/util/docker/docker_util.go]: Unfortunately, not anymore: I’m now using that function from `func (ku *KubeUtil) IsAgentHostNetwork()` in order to have the agent able to introspect the POD it is running in.

File: pkg/util/hostname.go
Comment (olivielpeau) [inline pkg/util/hostname.go]: Nit: if the call to `isOSHostnameUsable` either:

* has any chance to return inconsistent results between 2 calls (because of network calls/flakiness), or
* is "slow" (basically relies on network calls)

then for consistency and performance I think it'd be better to store this function's return value once at the beginning of the function

File: releasenotes/notes/check_hostname_is_usable-dc4b99e6f7242b34.yaml
Comment (olivielpeau) [inline releasenotes/notes/check_hostname_is_usable-dc4b99e6f7242b34.yaml]: Let's try to make this upgrade note as actionable as possible for users. I.e something like:

>    Starting with this version, the containerized Agent never chooses the OS hostname as its hostname when it is running in a dedicated UTS namespace. This is done in order to avoid picking container IDs or kubernetes POD names as hostnames, since these identifiers do not reflect the identity of the host they run on.
>
>    This change only affects you if [...] [here, add some additional info about how users can know if they are affected by the change, if the above info is not enough. And more importantly, write something about what users should do if they're affected by the change].

What do you think?

(if needed the PR can be merged as-is, and this release note can be changed later)

**Review by olivielpeau — DISMISSED** [review]
LGTM overall! Only reviewed the changes to `util/hostname.go` and the release note.

Left a small comment on the code, and another one on the release note. But no need for further approval from me before merging 👍

**Pull request merged**
